### PR TITLE
[GH-8471] Undeprecate PARTIAL for objects in DQL

### PR DESCRIPTION
**Pull request by beberlei**
With this PR SELECT PARTIAL p.{} is allowed again with object hydration. It was already re-allowed for array-hydration a few commits before.

The reason we want to allow this again is that PHP 8.4 Lazy Proxies will allow us to implement partial objects as proxies that will trigger lazy initialization as soon as one of the "not loaded" partial properties is accessed.

Related:
* https://github.com/doctrine/orm/issues/8471
* https://github.com/doctrine/orm/pull/11647

### CODE REVIEW

**Comment by Wirone** [comment]
Shouldn't it be "Revert deprecation of" or "Undeprecate" though?

File: docs/en/reference/partial-objects.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
+Partial Objects
+===============
+
+
+.. note::
+
+    Creating Partial Objects through DQL is deprecated and
+    will be removed in the future, use data transfer object
+    support in DQL instead. (\ `Details
+    <https://github.com/doctrine/orm/issues/8471>`_)
```
Comment (SenseException) [inline docs/en/reference/partial-objects.rst]:
```suggestion
```

I hope they're here to stay :smiley:

File: src/UnitOfWork.php [hunk 2]
```diff
@@ -2417,6 +2412,18 @@ public function createEntity(string $className, array $data, array &$hints = [])
...
+            Deprecation::trigger(
+                'doctrine/orm',
+                'https://github.com/doctrine/orm/issues/8471',
+                'Partial Objects are deprecated (here entity %s)',
+                $className,
+            );
+
+            return $entity;
+        }
+
```
Comment (SenseException) [inline src/UnitOfWork.php]:
```suggestion
```

Old deprecation message

**Review by SenseException — CHANGES_REQUESTED** [review]

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 3]
```diff
@@ -533,14 +533,23 @@ back. Instead, you receive only arrays as a flat rectangular result
...
+If you want to select partial objects or fields in array hydration you can use the ``partial``
+DQL keyword:
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT partial u.{id, username} FROM CmsUser u');
+    $users = $query->getResult(); // array of partially loaded CmsUser objects
+
+You use the partial syntax when joining as well:
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: I think a word is missing here.

```suggestion
You can use the partial syntax when joining as well:
```

File: docs/en/reference/partial-objects.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
+Partial Objects
+===============
+
+A partial object is an object whose state is not fully initialized
+after being reconstituted from the database and that is
+disconnected from the rest of its data. The following section will
+describe why partial objects are problematic and what the approach
+of Doctrine2 to this problem is.
```
Comment (greg0ire) [inline docs/en/reference/partial-objects.rst]: Should we keep using Doctrine2 here? I'm not sure our audience knows what Doctrine1 refers to.

File: docs/en/reference/partial-objects.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+These are reasons why many ORMs do not allow partial objects at all
+and instead you always have to load an object with all its fields
+(associations being proxied). One secure way to allow partial
+objects is if the programming language/platform allows the ORM tool
+to hook deeply into the object and instrument it in such a way that
+individual fields (not only associations) can be loaded lazily on
+first access. This is possible in Java, for example, through
+bytecode instrumentation. In PHP though this is not possible, so
+there is no way to have "secure" partial objects in an ORM with
+transparent persistence.
```
Comment (greg0ire) [inline docs/en/reference/partial-objects.rst:56]: Is this sentence still true? Or should we say it's not implemented yet?

File: docs/en/reference/partial-objects.rst (hunk 4 above)
Comment (SenseException) [reply docs/en/reference/partial-objects.rst]: We already had PRs where Doctrine2 was replaced with Doctrine. Let's do the same here.

Comment (beberlei) [inline docs/en/reference/partial-objects.rst]:
```suggestion
of Doctrine to this problem is.
```

File: docs/en/reference/partial-objects.rst (hunk 5 above)
Comment (beberlei) [reply docs/en/reference/partial-objects.rst:56]: Its still true until we get lazy objects.

**Review by SenseException — APPROVED** [review]

**Review by greg0ire — APPROVED** [review]

**Pull request merged**